savant/base/pyfunc.py

```diff
@@ -70,6 +71,34 @@ def process_buffer(self, buffer: Gst.Buffer):
         :param buffer: Gstreamer buffer.
         """
 
+    def get_queues(self) -> List[Gst.Element]:
+        """Get upstream queues.
+
+        Can be used to check the current queue length for bandwidth management.
+
+        E.g.:
+        ```
+        for queue in self.get_queues():
+            queue_length = queue.get_property('current-level-buffers')
+            if queue_length > 10:
+                self.logger.warning('Queue %s is full!', queue.get_name())
+        ```
+        """
+        if self.gst_element is not None:
+            gst_pipeline = self.gst_element.get_parent()
+            return get_elements(
+                gst_pipeline,
+                factory='queue',
+                before_element=self.gst_element,
+            )
+        return []
+
+    def get_queue(self, name: str) -> Optional[Gst.Element]:
```

Comment:
@denisvmedyantsev, this implementation is probably suboptimal. Does GStreamer not allow getting exact elements by name? If not, do we need to keep a separate reference table for them? I do not like that we iterate over elements just to find the required. Or, at least cache them in order to avoid excessive iterations.

I also do not like the exposing Gst.Element as a resulting type. If we return a queue, then we need a type representing a queue.

Reply:
Yes, we can use `get_by_name`, I'll change the implementation.

"I also do not like the exposing Gst.Element as a resulting type. If we return a queue, then we need a type representing a queue." - python doesn't have Gst.Queue, it's not exactly a type... I don't think we have something to replace Gst.Element here.

Reply:
Changed get_queue(name) implementation.

Add GstQueue wrapper for queue element to simplify access to main properties.